### PR TITLE
B 20621 mobile shipment apis int 2

### PR DESCRIPTION
**Pull request by KonstanceH**
## Summary

This is to add/modify services and Internal api endpoints related to Mobile Home Shipment on customer's side of the app.

- Added Mobile Home Shipment services(updater, fetcher, creator)
- Updated MTOShipment services to include mobile home shipment.
- Added mobile home shipment to MTOShipment internal API(GET, POST, PATCH).
- Feature Flag added to Internal, GHC, and Prime MTOShipment  API Handlers(GET).

### How to test
1. Create a move with any shipment.
2. As a customer ->  click "Go to Move" for the current move.
3. Use the browser dev tools get the info you need for the Swagger or Postman.

Use Swagger or Postman to check Internal APIs:

Import `internal.yaml` file from swagger folder if you have not already.

POST - http://milmovelocal:3000/swagger-ui/internal.html#/mtoShipment/createMTOShipment
Example body to use:
```
{
  "moveTaskOrderID": "a615abd5-c64a-4c0d-b438-2b5e66790130",
  "shipmentType": "MOBILE_HOME",
  "mobileHomeShipment": {
    "year": 1111,
    "model": "Mobile Home Model",
    "make": "Mobile Home Make",
    "lengthInInches": 169,
    "widthInInches": 83,
    "heightInInches": 78,
  }
}
```

PATCH - http://milmovelocal:3000/swagger-ui/internal.html#/mtoShipment/updateMTOShipment
Example body to use:
```
{
  "moveTaskOrderID": "2ceb3116-b24b-41df-a023-67ac09ec13de",
  "shipmentType": "MOBILE_HOME",
  "mobileHomeShipment": {
    "model": "Mobile Home Model 2"
  }
}
```

GET - http://milmovelocal:3000/swagger-ui/internal.html#/mtoShipment/listMTOShipments

GET - http://milmovelocal:3000/swagger-ui/internal.html#/moves/getAllMoves

### Backend

- [ ] Code follows the guidelines for [Logging](https://transcom.github.io/mymove-docs/docs/getting-started/development/logging).
- [ ] The requirements listed in [Querying the Database Safely](https://transcom.github.io/mymove-docs/docs/backend/guides/golang-guide#querying-the-database-safely) have been satisfied.

#### Any new migrations/schema changes:

- [ ] Follows our guidelines for [Zero-Downtime Deploys](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#zero-downtime-migrations).
- [ ] Have been communicated to #g-database.
- [ ] Secure migrations have been tested following the instructions in our [docs](https://transcom.github.io/mymove-docs/docs/backend/setup/database-migrations#secure-migrations).

## Screenshots
_Grab moveTaskOrder ID from developer tools_
Screenshot 2024-08-13 at 11 19 04 AM

_Example of swagger input for mobile home POST_
Screenshot 2024-08-13 at 11 18 13 AM

_Mobile Home model returned after POST_
Screenshot 2024-08-13 at 11 18 05 AM

_Mobile Home model returned after PATCH_
Screenshot 2024-08-13 at 11 59 41 AM

### CODE REVIEW

**Review by taeJungCaci — COMMENTED** [review]
Missing Feature flag in `config/env/loadtest.app.env`

File: pkg/handlers/ghcapi/mto_shipment.go [hunk 1]
```diff
@@ -89,6 +89,35 @@ func (h ListMTOShipmentsHandler) Handle(params mtoshipmentops.ListMTOShipmentsPa
 			}
 			/** End of Feature Flag **/
 
+			/** Feature Flag - Mobile Home Shipment **/
+			// TODO: check if name change is necessary
+			featureFlagNameMH := "mobileHome"
```
Comment (taeJungCaci) [inline pkg/handlers/ghcapi/mto_shipment.go:94]: This might have to be `mobile_home`

File: pkg/handlers/internalapi/moves.go [hunk 2]
```diff
@@ -387,6 +387,35 @@ func (h GetAllMovesHandler) Handle(params moveop.GetAllMovesParams) middleware.R
 				}
 				/** End of Feature Flag Block **/
 
+				/** Feature Flag - Mobile Home Shipment **/
+				// TODO: check if name change is necessary
+				featureFlagNameMH := "mobileHome"
```
Comment (taeJungCaci) [inline pkg/handlers/internalapi/moves.go:392]: same with here.

File: pkg/handlers/internalapi/mto_shipment.go [hunk 3]
```diff
@@ -293,6 +293,35 @@ func (h ListMTOShipmentsHandler) Handle(params mtoshipmentops.ListMTOShipmentsPa
 			}
 			/** End of Feature Flag **/
 
+			/** Feature Flag - Mobile Home Shipment **/
+			// TODO: check if name change is necessary
+			featureFlagNameMH := "mobileHome"
```
Comment (taeJungCaci) [inline pkg/handlers/internalapi/mto_shipment.go:298]: same

File: pkg/handlers/primeapi/move_task_order.go [hunk 4]
```diff
@@ -115,6 +115,34 @@ func (h GetMoveTaskOrderHandler) Handle(params movetaskorderops.GetMoveTaskOrder
 			}
 			/** End of Feature Flag **/
 
+			/** Feature Flag - Mobile Home Shipment **/
+			isMobileHomeFeatureOn := false
+			featureFlagNameMH := "mobileHome"
```
Comment (taeJungCaci) [inline pkg/handlers/primeapi/move_task_order.go:120]: same

File: pkg/handlers/primeapiv2/move_task_order.go [hunk 5]
```diff
@@ -78,6 +78,34 @@ func (h GetMoveTaskOrderHandler) Handle(params movetaskorderops.GetMoveTaskOrder
 			}
 			/** End of Feature Flag **/
 
+			/** Feature Flag - Mobile Home Shipment **/
+			isMobileHomeFeatureOn := false
+			featureFlagNameMH := "mobileHome"
```
Comment (taeJungCaci) [inline pkg/handlers/primeapiv2/move_task_order.go:83]: same

File: pkg/handlers/primeapiv3/move_task_order.go [hunk 6]
```diff
@@ -78,6 +78,34 @@ func (h GetMoveTaskOrderHandler) Handle(params movetaskorderops.GetMoveTaskOrder
 			}
 			/** End of Feature Flag **/
 
+			/** Feature Flag - Mobile Home Shipment **/
+			isMobileHomeFeatureOn := false
+			featureFlagNameMH := "mobileHome"
```
Comment (taeJungCaci) [inline pkg/handlers/primeapiv3/move_task_order.go:83]: same

File: pkg/services/mobile_home_shipment/rules.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+			verrs.Add("widthInInches", "cannot be a zero or a negative value")
+		}
+		if newMobileHomeShipment.HeightInInches == nil || *newMobileHomeShipment.HeightInInches <= 0 {
+			verrs.Add("heightInInches", "cannot be a zero or a negative value")
+		}
+		if newMobileHomeShipment.HeightInInches != nil && newMobileHomeShipment.LengthInInches != nil && newMobileHomeShipment.WidthInInches != nil {
+			if *newMobileHomeShipment.LengthInInches <= 168 && *newMobileHomeShipment.WidthInInches <= 82 && *newMobileHomeShipment.HeightInInches <= 77 {
+				verrs.Add("heightInInches", "One of these criteria must be met for it to be a mobile home shipment: lengthInInches > 168, widthInInches > 82, or heightInInches > 77.")
+				verrs.Add("widthInInches", "One of these criteria must be met for it to be a mobile home shipment: lengthInInches > 168, widthInInches > 82, or heightInInches > 77.")
+				verrs.Add("lengthInInches", "One of these criteria must be met for it to be a mobile home shipment: lengthInInches > 168, widthInInches > 82, or heightInInches > 77.")
```
Comment (taeJungCaci) [inline pkg/services/mobile_home_shipment/rules.go:72]: I think the validations for mobile home is different. Might want to check if there is one.

**Review by taeJungCaci — CHANGES_REQUESTED** [review]

File: migrations/app/schema/20240813153431_change_mobile_home_shipment_deleted_at_type.up.sql [hunk 8]
```diff
@@ -0,0 +1 @@
+ALTER TABLE mobile_homes ALTER COLUMN deleted_at drop not NULL
```
[the remark this answers is not in the document]
Comment (taeJungCaci) [reply migrations/app/schema/20240813153431_change_mobile_home_shipment_deleted_at_type.up.sql:1]: and don't forget to make these changes from your Main branch plz

**Review by taeJungCaci — CHANGES_REQUESTED** [review]
I think you may also have to add mobile home to `pkg/models/mto_shipments.go` model

File: pkg/services/mto_service_item/mto_service_item_updater.go
Comment (taeJungCaci) [inline pkg/services/mto_service_item/mto_service_item_updater.go:1]: I think we want to keep these changes - I don't see these changes in the revert PR.

File: pkg/services/mto_service_item/mto_service_item_updater_test.go
Comment (taeJungCaci) [inline pkg/services/mto_service_item/mto_service_item_updater_test.go:1]: Same with these

File: pkg/services/mto_service_item/mto_service_item_validators.go
Comment (taeJungCaci) [inline pkg/services/mto_service_item/mto_service_item_validators.go:1]: same with these

File: pkg/services/mto_service_item/mto_service_item_validators_test.go
Comment (taeJungCaci) [inline pkg/services/mto_service_item/mto_service_item_validators_test.go:1]: same with these

File: playwright/tests/office/servicescounseling/servicesCounselingNTS.spec.js
Comment (taeJungCaci) [inline playwright/tests/office/servicescounseling/servicesCounselingNTS.spec.js:1]: same with these

File: playwright/tests/office/txo/tooFlowsNTS.spec.js
Comment (taeJungCaci) [inline playwright/tests/office/txo/tooFlowsNTS.spec.js:1]: same with these

**Review by taeJungCaci — CHANGES_REQUESTED** [review]

File: migrations/app/migrations_manifest.txt [hunk 9]
```diff
@@ -974,12 +974,12 @@
 20240729162353_joseph_doye_cn_cac.up.sql
 20240729164930_mai_do_cac.up.sql
 20240729185147_add_cancel_to_ppm_status_enum.up.sql
-20240729200347_add_mto_approved_at_timestamp.up.sql
 20240730161630_remove-boat-shipments-index.up.sql
-20240731125005_retroactively_update_approve_at_column_based_on_available_to_prime.up.sql
```
Comment (taeJungCaci) [inline migrations/app/migrations_manifest.txt:979]: I think the plan might be to keep these in - they added a new migration file to remove these

**Review by taeJungCaci — CHANGES_REQUESTED** [review]

**Comment by KonstanceH** [comment]
> pkg/models/mto_shipments.go

I have it added in that file from the Backend/DB PR.

**Review by taeJungCaci — APPROVED** [review]
LGTM!

**Review by taeJungCaci — DISMISSED** [review]
I forgot to add - could you add a `?` to `switch (shipmentType?.toUpperCase()) {` line 185 in ShipmentList.jsx please? It is something I've missed for mine and the app errors out if I create a shipment and try to view it in the UI. I know UI does not matter for this story, but I think for testing/POA purposes, they may want to see it populate and not crash.

**Review by brooklyn-welsh — APPROVED** [review]
Looks like all the changes were made and went through okay, LGTM!

**Review by taeJungCaci — APPROVED** [review]

**Review by taeJungCaci — COMMENTED** [review]
I didn't wanna stop the PR for this, but just FYI - Adding a `?` to `switch (shipmentType?.toUpperCase()) { ` line 185 in ShipmentList.jsx  will stop the customer app from crashing when getting shipments after doing creating a mobile shipment.

**Comment by KonstanceH** [comment]
> I didn't wanna stop the PR for this, but just FYI - Adding a `?` to `switch (shipmentType?.toUpperCase()) { ` line 185 in ShipmentList.jsx will stop the customer app from crashing when getting shipments after doing creating a mobile shipment.

ah okay. That's no problem since I can't merge into integration now anyway haha. I'll make the change and push it up now

**Pull request merged**